### PR TITLE
8297503: Merge FilteringDCTOC into ContiguousSpaceDCTOC

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple cleanup of trimming unnecessary type hierarchy levels.

Test: hotspot_gc

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297503](https://bugs.openjdk.org/browse/JDK-8297503): Merge FilteringDCTOC into ContiguousSpaceDCTOC

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11322/head:pull/11322` \
`$ git checkout pull/11322`

Update a local copy of the PR: \
`$ git checkout pull/11322` \
`$ git pull https://git.openjdk.org/jdk pull/11322/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11322`

View PR using the GUI difftool: \
`$ git pr show -t 11322`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11322.diff">https://git.openjdk.org/jdk/pull/11322.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shared/space.hpp [hunk 1]
```diff
@@ -541,63 +541,42 @@ class ContiguousSpace: public CompactibleSpace {
...
 };
 
-
-// A dirty card to oop closure that does filtering.
-// It knows how to filter out objects that are outside of the _boundary.
-class FilteringDCTOC : public DirtyCardToOopClosure {
+// A dirty card to oop closure for contiguous spaces (ContiguousSpace and
+// sub-classes). It knows how to filter out objects that are outside of the
+// _boundary.
+//
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/space.hpp:547]:
```suggestion
//
// Assumptions:
```

File: src/hotspot/share/gc/shared/space.hpp [hunk 2]
```diff
@@ -541,63 +541,42 @@ class ContiguousSpace: public CompactibleSpace {
...
-                                       OopIterateClosure* cl);
-  virtual void walk_mem_region_with_cl(MemRegion mr,
-                                       HeapWord* bottom, HeapWord* top,
-                                       FilteringClosure* cl);
+  void walk_mem_region_with_cl(MemRegion mr,
+                               HeapWord* bottom, HeapWord* top,
+                               OopIterateClosure* cl);
+  void walk_mem_region_with_cl(MemRegion mr,
+                               HeapWord* bottom, HeapWord* top,
+                               FilteringClosure* cl);
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/space.hpp:573]: These do not need to be `protected` afaict

File: src/hotspot/share/gc/shared/space.hpp [hunk 3]
```diff
@@ -541,63 +541,42 @@ class ContiguousSpace: public CompactibleSpace {
...
 
 public:
   ContiguousSpaceDCTOC(ContiguousSpace* sp, OopIterateClosure* cl,
                        CardTable::PrecisionStyle precision,
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/space.hpp:577]: That's a pre-existing (different) issue, but the only `PrecisionStyle` this code is ever used with seems to be `ObjHeadPreciseArray`.

File: src/hotspot/share/gc/shared/space.hpp [hunk 4]
```diff
@@ -541,63 +541,42 @@ class ContiguousSpace: public CompactibleSpace {
...
+//
+// 1. That the actual top of any area in a memory region
+//    contained by the space is bounded by the end of the contiguous
+//    region of the space.
+// 2. That the space is really made up of objects and not just
+//    blocks.
+class ContiguousSpaceDCTOC : public DirtyCardToOopClosure {
 protected:
-  // Override.
+  // Overrides.
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/space.hpp:555]: Instead of the `overrides` comment, both methods could be decorated with the `override` specifier.

**Review by tschatzl — APPROVED** [review]
Lgtm, some minor comments to be fixed as you want.

File: src/hotspot/share/gc/shared/space.hpp (hunk 3 above)
Comment (albertnetymk) [reply src/hotspot/share/gc/shared/space.hpp:577]: Filed at https://bugs.openjdk.org/browse/JDK-8297572

File: src/hotspot/share/gc/shared/space.hpp (hunk 4 above)
Comment (albertnetymk) [reply src/hotspot/share/gc/shared/space.hpp:555]: I will do that in another PR for the complete file.

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**